### PR TITLE
[v4] Add support for client data telemetry with CLI_DATA parameter

### PR DESCRIPTION
**Pull request by konstantin-msft**
This pull request introduces a new mechanism for parsing and instrumenting the `clientdata` parameter returned from the `/authorize` endpoint, enhancing telemetry and error reporting for authentication flows. The main changes include the addition of a robust parser for the `clientdata` field, its integration into the telemetry pipeline, and comprehensive tests to ensure correct parsing and instrumentation. API review files and constants are also updated to reflect these changes.

**Clientdata Parsing and Instrumentation:**

- Added a new `parseClientData` function to extract structured information (account type, error, suberror, cloud instance, and caller data boundary) from the `clientdata` parameter, with robust handling for malformed or missing data. (`lib/msal-browser/src/protocol/Authorize.ts`, `lib/msal-browser/test/protocol/ClientDataParser.spec.ts`) [[1]](diffhunk://#diff-a4a9623518687b3ceac75fec2a3748ba77e40fd2166581e3ff0c4149258c6165R48-R125) [[2]](diffhunk://#diff-5d9629a97f66d7059e7e9ba3cc5efcd1a65407d6c346145d19d1a46854705876R1-R130)
- Introduced the `instrumentClientData` function, which processes the parsed `clientdata` and instruments telemetry fields (`accountType`, `serverErrorNo`, `serverSubErrorNo`) via the `performanceClient`. This function is now called in both `handleResponseCode` and `handleResponseEAR` flows to ensure telemetry is captured early. (`lib/msal-browser/src/protocol/Authorize.ts`) [[1]](diffhunk://#diff-a4a9623518687b3ceac75fec2a3748ba77e40fd2166581e3ff0c4149258c6165R493-R496) [[2]](diffhunk://#diff-a4a9623518687b3ceac75fec2a3748ba77e40fd2166581e3ff0c4149258c6165R577-R579)

**API and Telemetry Updates:**

- Updated API review files to add the `CLI_DATA` constant, the `addCliData` function, and the `clientdata` property to the `AuthorizeResponse` type, reflecting the new parameter and its usage. Also added `serverSubErrorNo` to the `PerformanceEvent` type. (`lib/msal-common/apiReview/msal-common.api.md`) [[1]](diffhunk://#diff-09087b913ebbfa828e5f36b7476a400328e0a7131db84f622cc5f6994759a117L81-R82) [[2]](diffhunk://#diff-09087b913ebbfa828e5f36b7476a400328e0a7131db84f622cc5f6994759a117R272-R276) [[3]](diffhunk://#diff-09087b913ebbfa828e5f36b7476a400328e0a7131db84f622cc5f6994759a117R840) [[4]](diffhunk://#diff-09087b913ebbfa828e5f36b7476a400328e0a7131db84f622cc5f6994759a117R1349-R1353) [[5]](diffhunk://#diff-09087b913ebbfa828e5f36b7476a400328e0a7131db84f622cc5f6994759a117R3458) [[6]](diffhunk://#diff-09087b913ebbfa828e5f36b7476a400328e0a7131db84f622cc5f6994759a117R3909)

**Testing Enhancements:**

- Added a dedicated test suite for `parseClientData`, covering various edge cases, malformed input, and correct parsing for both MSA and Entra (AAD) account types. (`lib/msal-browser/test/protocol/ClientDataParser.spec.ts`)
- Expanded the `Authorize` protocol tests to verify that `clidata=1` is included in requests, and that telemetry fields are correctly instrumented from `clientdata` in both success and error scenarios. (`lib/msal-browser/test/protocol/Authorize.spec.ts`) [[1]](diffhunk://#diff-383f979b9a05a2d7fe02052138e8087f6ca2167e78d44dc4d41c391463d4da04R197) [[2]](diffhunk://#diff-383f979b9a05a2d7fe02052138e8087f6ca2167e78d44dc4d41c391463d4da04R470-R725)

These changes collectively improve error diagnostics, telemetry, and maintainability of the authentication flows by providing structured, early insight into server responses.

### CODE REVIEW

File: lib/msal-browser/src/protocol/Authorize.ts [hunk 1]
```diff
@@ -45,6 +45,84 @@ import { EventHandler } from "../event/EventHandler.js";
...
+    if (parsed) {
+        performanceClient.addFields(
+            {
+                accountType: parsed.accountType,
+                serverErrorNo: parsed.error,
+                serverSubErrorNo: parsed.subError,
+            },
+            correlationId
+        );
+    }
```
[the remark this answers is not in the document]
Comment (konstantin-msft) [reply lib/msal-browser/src/protocol/Authorize.ts]: fixed

File: change/@azure-msal-common-17d28873-a06b-4a2d-b3c0-9a775064086b.json [hunk 2]
```diff
@@ -0,0 +1,7 @@
+{
+  "type": "minor",
+  "comment": "Add support for client data telemetry with CLI_DATA parameter #8378",
```
[the remark this answers is not in the document]
Comment (konstantin-msft) [reply change/@azure-msal-common-17d28873-a06b-4a2d-b3c0-9a775064086b.json]: fixed

File: lib/msal-browser/test/protocol/ClientDataParser.spec.ts [hunk 3]
```diff
@@ -0,0 +1,130 @@
+import { parseClientData } from "../../src/protocol/Authorize.js";
+
+describe("parseAuxClientData", () => {
+    describe("parseAuxClientData", () => {
+        it("returns null for undefined input", () => {
```
[the remark this answers is not in the document]
Comment (konstantin-msft) [reply lib/msal-browser/test/protocol/ClientDataParser.spec.ts]: fixed

File: lib/msal-browser/src/protocol/Authorize.ts [hunk 4]
```diff
@@ -45,6 +45,84 @@ import { EventHandler } from "../event/EventHandler.js";
...
+            return null;
+        }
+
+        return {
+            accountType: clientDataAccountTypeMapping.get(parts[0] || "") || "",
+            error: parts[1]?.trim() || "",
+            subError: parts[2]?.trim() || "",
+            cloudInstance: parts[3]?.trim() || "",
+            callerDataBoundary: parts[4]?.trim() || "",
+        };
```
[the remark this answers is not in the document]
Comment (konstantin-msft) [reply lib/msal-browser/src/protocol/Authorize.ts:100]: fixed

File: lib/msal-browser/src/protocol/Authorize.ts [hunk 5]
```diff
@@ -45,6 +45,84 @@ import { EventHandler } from "../event/EventHandler.js";
...
 
+/**
+ * Parsed representation of the clientdata response parameter from the /authorize endpoint.
+ *
+ * Format: urlencoded(account_type|error|sub_error|cloud_instance|caller_data_boundary)
+ */
+type ClientData = {
+    /** Account type: "m" for MSA, "e" for Entra */
+    accountType: string;
+    /** Error code string (e.g. "0x8004345C" for MSA) */
```
[the remark this answers is not in the document]
Comment (konstantin-msft) [reply lib/msal-browser/src/protocol/Authorize.ts:56]: fixed

File: change/@azure-msal-browser-ae790b79-88ea-45d8-929a-1373ad2b9f6c.json [hunk 6]
```diff
@@ -0,0 +1,7 @@
+{
+  "type": "minor",
+  "comment": "Add support for client data telemetry with CLI_DATA parameter #8378",
```
[the remark this answers is not in the document]
Comment (konstantin-msft) [reply change/@azure-msal-browser-ae790b79-88ea-45d8-929a-1373ad2b9f6c.json]: fixed

**Review by peterzenz — DISMISSED** [review]

**Review by peterzenz — DISMISSED** [review]

File: lib/msal-browser/src/protocol/Authorize.ts [hunk 7]
```diff
@@ -45,6 +45,90 @@ import { EventHandler } from "../event/EventHandler.js";
...
 import { IPlatformAuthHandler } from "../broker/nativeBroker/IPlatformAuthHandler.js";
 
+/**
+ * Parsed representation of the clientdata response parameter from the /authorize endpoint.
+ *
+ * Format: urlencoded(account_type|error|sub_error|cloud_instance|caller_data_boundary)
+ */
+type ClientData = {
+    /** Account type: MSA, AAD */
+    accountType: string;
```
Comment (sameerag) [inline lib/msal-browser/src/protocol/Authorize.ts:55]: What is the plan to identify AAD? If we do not have this in the response, and we know the authority we map it?

File: lib/msal-browser/src/protocol/Authorize.ts [hunk 8]
```diff
@@ -45,6 +45,90 @@ import { EventHandler } from "../event/EventHandler.js";
...
+    /** Sub-error code string (e.g. "0x80043588" for MSA) */
+    subError: string;
+    /** Cloud instance hostname (e.g. "login.microsoftonline.com") */
+    cloudInstance: string;
+    /** Caller data boundary (e.g. "none" for MSA) */
+    callerDataBoundary: string;
+};
+
+const clientDataAccountTypeMapping = new Map([
+    ["e", "AAD"],
```
Comment (sameerag) [inline lib/msal-browser/src/protocol/Authorize.ts:67]: I do not expect AAD set for every request, or am I missing something?

**Review by sameerag — DISMISSED** [review]
Approved. Comments about the data content we receive, non blocking.

File: lib/msal-browser/src/protocol/Authorize.ts (hunk 7 above)
Comment (konstantin-msft) [reply lib/msal-browser/src/protocol/Authorize.ts:55]: Account object in auth response takes precedence over this one, which we use as a fallback in case of an error.

File: lib/msal-browser/src/protocol/Authorize.ts (hunk 8 above)
Comment (konstantin-msft) [reply lib/msal-browser/src/protocol/Authorize.ts:67]: Not sure i follow. It's AAD, MSA or empty

Comment (sameerag) [reply lib/msal-browser/src/protocol/Authorize.ts:67]: Never mind, I got it I think with your previous response. Thanks. Resolving.

**Review by sameerag — APPROVED** [review]

**Review by peterzenz — APPROVED** [review]
